### ISSUE TITLE
Linear regression

### ISSUE THREAD

**Issue by JohanGarp** [Microsoft/LightGBM#917]
I have the impression that LightGBM regression means logistic regression rather than linear regression,
is that correct? If so, are there any plans to support linear regression?

**Comment by Laurae2** [comment]
Binary objective is logistic regression, while regression (same as regression_l2) objective is like any linear regression with least squares through gradient descent.

**Comment by JohanGarp** [comment]
That was quick response!
Thank you, I will try that immediately!

**Comment by JohanGarp** [comment]
Maybe there is something that I don't understand, but here is a simple example in code:
  for (int i = 0; i < 200; ++i)
  {
    for (int j = 0; j < 200; ++j)
    {
      cout << i+j << ", " << i << ", " << j << endl;
    }
  }

This should generate values from 0 to 400.
I use the same data for training and testing.
Output from LightGBM is values from 194 to 203.

**Comment by Laurae2** [comment]
This is normal, it is a tree-based model. It is not a linear model, but a non-linear model made of thresholds as rules.

**Comment by JohanGarp** [comment]
OK, thanks.

**Comment by Laurae2** [comment]
@guolinke How do you make @JohanGarp example work *properly* while being reproducible? Here is an example using an integer label from 1 to 200 using features with the same exact values:

```r
lgb.unloader(wipe = TRUE)
library(lightgbm)
train_data <- matrix(1:200, ncol = 1)
train_label <- matrix(1:200, ncol = 1)
dtrain <- lgb.Dataset(train_data, label = train_label)
valids <- list(train = dtrain)
params <- list(objective = "regression",
               metric = "l2",
               min_data = 1,
               min_data_in_bin = 5,
               min_hessian = 1e-3,
               num_threads = 1)
model <- lgb.train(params,
                   dtrain,
                   5,
                   valids,
                   learning_rate = 1,
                   early_stopping_rounds = 10)
plot(predict(model, train_data))
```

And the weird binning result:

![image](https://user-images.githubusercontent.com/9083669/30495807-b0ec3110-9a4d-11e7-86db-604f425360ec.png)

**Comment by guolinke** [comment]
@Laurae2 try the min_data_in_bin=1, and bigger num_leaves/num_iterations

**Comment by Laurae2** [comment]
It refuses to train with too large number of leaves (same if min_data_in_bin is lowered to 1).

**Comment by guolinke** [comment]
@Laurae2 refuses ?

**Comment by Laurae2** [comment]
@guolinke Try this:

```r
lgb.unloader(wipe = TRUE)
library(lightgbm)
train_data <- matrix(1:200, ncol = 1)
train_label <- matrix(1:200, ncol = 1)
dtrain <- lgb.Dataset(train_data, label = train_label)
valids <- list(train = dtrain)
params <- list(objective = "regression",
               metric = "l2",
               min_data = 1,
               min_data_in_bin = 1,
               min_hessian = 1e-3,
               num_threads = 1,
               num_leaves = 255)
model <- lgb.train(params,
                   dtrain,
                   5,
                   valids,
                   learning_rate = 1,
                   early_stopping_rounds = 10)
plot(predict(model, train_data))
```

You will get an error if you try running it multiple times. The training also would be very poor.

The error:

```r
Error in lgb.call("LGBM_DatasetCreateFromMat_R", ret = handle, private$raw_data,  : 
  api error: Cannot construct Dataset since there are not useful features.
```

![image](https://user-images.githubusercontent.com/9083669/30512979-78b67702-9afb-11e7-8697-917f71aa0e05.png)

**Comment by guolinke** [comment]
@Laurae2 
test it in the CLI:

log:

```
C:\Users\xxx\Documents\LightGBM\windows\x64\Debug>lightgbm.exe data=test.txt min_data=1 min_data_in_bin=1 num_leaves=255 metric=l2 test=test.txt learning_rate=1 boost_from_average=false num_iterations=10
[LightGBM] [Info] Finished loading parameters
[LightGBM] [Info] Finished loading data in 0.046692 seconds
[LightGBM] [Info] Total Bins 201
[LightGBM] [Info] Number of data: 200, number of used features: 1
[LightGBM] [Info] Finished initializing training
[LightGBM] [Info] Started training...
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] Iteration:1, valid_1 l2 : 1691.75
[LightGBM] [Info] 0.014806 seconds elapsed, finished iteration 1
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] Iteration:2, valid_1 l2 : 2.1016e-27
[LightGBM] [Info] 0.029382 seconds elapsed, finished iteration 2
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] Iteration:3, valid_1 l2 : 0
[LightGBM] [Info] 0.035889 seconds elapsed, finished iteration 3
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements.
[LightGBM] [Info] 0.042311 seconds elapsed, finished iteration 4
[LightGBM] [Info] Finished training
```

data: 
[test.txt](https://github.com/Microsoft/LightGBM/files/1308417/test.txt)

**Comment by Laurae2** [comment]
@guolinke OK, found why it was not good: integers instead of numeric.

Code below works fine:

```r
lgb.unloader(wipe = TRUE)
library(lightgbm)
train_data <- matrix(as.numeric(1:200), ncol = 1)
train_label <- matrix(as.numeric(1:200), ncol = 1)
dtrain <- lgb.Dataset(train_data, label = train_label)
valids <- list(train = dtrain)
params <- list(objective = "regression",
               metric = "l2",
               min_data = 1,
               min_data_in_bin = 1,
               num_threads = 1,
               num_leaves = 255)
model <- lgb.train(params,
                   dtrain,
                   2,
                   valids,
                   learning_rate = 1,
                   early_stopping_rounds = 10)
plot(predict(model, train_data))
```

![image](https://user-images.githubusercontent.com/9083669/30513527-0b7145d2-9b05-11e7-9e64-4d8a240b3e5a.png)

**Comment by guolinke** [comment]
@Laurae2 
In the LightGBM R package, the dataframe is passed to lightgbm as a float array. When the dataframe is integer type, I cannot sure it will convert to float or not.

**Comment by JohanGarp** [comment]
I've used the CLI stuff in examples/regression as a starting point.
When looking at the parameters in your example, I noticed that I can increase the 
"learning_rate" from 0.05 to 1.
I also found a parameter "feature_fraction" that is not helping in my case :-)
Now I think it looks much better!

**Comment by vijaykumarnadal** [comment]
hai i am new to lightgbm package i have the following sample dataset(train) i am trying to build linear regression model with medianhousevalue(response variable) in rstudio

housingMedianAge | totalRooms | totalBedrooms | population | households | medianIncome | medianHouseValue
41 | 880 | 129 | 322 | 126 | 8.3252 | 452600
21 | 7099 | 1106 | 2401 | 1138 | 8.3014 | 358500
52 | 1467 | 190 | 496 | 177 | 7.2574 | 352100
52 | 1274 | 235 | 558 | 219 | 5.6431 | 341300
52 | 1627 | 280 | 565 | 259 | 3.8462 | 342200
52 | 919 | 213 | 413 | 193 | 4.0368 | 269700
52 | 2535 | 489 | 1094 | 514 | 3.6591 | 299200
52 | 3104 | 687 | 1157 | 647 | 3.12 | 241400
42 | 2555 | 665 | 1206 | 595 | 2.0804 | 226700
52 | 3549 | 707 | 1551 | 714 | 3.6912 | 261100

and the sample testset is as follows


52 | 2378 | 508 | 940 | 451 | 2.9583
48 | 2123 | 494 | 859 | 474 | 1.8523
50 | 1760 | 341 | 741 | 316 | 4.5
51 | 1262 | 309 | 608 | 298 | 2.1974
33 | 2821 | 652 | 1206 | 640 | 2.5481

can you pls share the code for building the model and predicting the output with params list.
